Support host and ref options in Menu.open

ianthomas23 opened the pull request:
Implements enhancement #699.

This exposes the `Widget.attach` optional arguments `host` and `ref` in `Menu.IOptions` to provide greater control over exactly where in the DOM the menu `div` is attached.

I have kept the names and descriptions of the options the same as in `Widget.attach`.

krassowski commented on packages/widgets/src/menu.ts:
```diff
@@ -1535,7 +1552,9 @@ namespace Private {
+    ref: HTMLElement | null
   ): void {
     // Get the current position and size of the main viewport.
     const windowData = getWindowData();
```
With the proposed changes the menu could end up beyond the boundaries of the element. This will be potentially problematic if the parent element (host) has clips the menu off. Is my thinking correct?

I think this is fine to merge as-is, I just wonder if this should be documented.

krassowski commented:
You will need to run `yarn run api` and commit changes.

fcollonval reviewed (approved):
Thanks @ianthomas23

FYI (not requiring code changes) it is better to used the non nullish operator `??` rather than the or operator.
The reason or is used a lot in lumino is that the code predates the introduction of that operator.

The goal is to avoid the case of a falsy value being overridden by the default like `0 || 1` != `0 ?? 1` (the first equals one when the second equals 0).

fcollonval commented:
> in addition to the API update, you will need to lint the code with `yarn lint`. One of the tests is failing due to

[warn] packages/widgets/tests/src/menu.spec.ts

ianthomas23 commented:
Thanks for the guidance @krassowski and @fcollonval. I've added the extra documentation to `Menu.open`, although it could possibly be added to `IOpenOptions` instead, as that was my entry point into the docs. I've also switched the `||` to `??` in the new code whilst I was there.

ianthomas23 commented:
I cannot tell if the failing `macos-latest` `webkit-headless` CI run is something that I have caused or if it is unrelated.

fcollonval commented:
> I cannot tell if the failing `macos-latest` `webkit-headless` CI run is something that I have caused or if it is unrelated.

This is due to flakiness - I let @krassowski have a second look.

fcollonval commented:
gentle ping to @krassowski

krassowski reviewed (approved):
Thanks!
